CesiumGeometry/src/IntersectionTests.cpp

```diff
@@ -100,6 +106,209 @@ std::optional<glm::dvec2> IntersectionTests::rayEllipsoid(
+  const glm::dvec3& direction = ray.getDirection();
+
+  glm::dvec3 edge0 = p1 - p0;
+  glm::dvec3 edge1 = p2 - p0;
+
+  glm::dvec3 p = glm::cross(direction, edge1);
+  double det = glm::dot(edge0, p);
+  if (cullBackFaces) {
+    if (det < Math::Epsilon6)
+      return false;
+
+    glm::dvec3 tvec = origin - p0;
+    double u = glm::dot(tvec, p);
+    if (u < 0.0 || u > det)
+      return false;
+
+    glm::dvec3 q = glm::cross(tvec, edge0);
+    double v = glm::dot(direction, q);
+    if (v < 0.0 || u + v > det)
+      return false;
+
+    t = glm::dot(edge1, q) / det;
+    return true;
+
+  } else {
+
+    if (glm::abs(det) < Math::Epsilon6)
+      return false;
+
+    double invDet = 1.0 / det;
+
+    glm::dvec3 tvec = origin - p0;
+    double u = glm::dot(tvec, p) * invDet;
+    if (u < 0.0 || u > 1.0)
+      return false;
+
+    glm::dvec3 q = glm::cross(tvec, edge0);
+    double v = glm::dot(direction, q) * invDet;
+    if (v < 0.0 || u + v > 1.0)
+      return false;
+
+    t = glm::dot(edge1, q) * invDet;
+    return true;
+  }
+}
+
+std::optional<glm::dvec3>
+IntersectionTests::rayAABB(const Ray& ray, const AxisAlignedBox& aabb) {
+  double t;
+  return rayAABBParametric(ray, aabb, t) && t >= 0
+             ? std::make_optional<glm::dvec3>(ray.pointFromDistance(t))
+             : std::nullopt;
+}
+
+bool IntersectionTests::rayAABBParametric(
+    const Ray& ray,
+    const AxisAlignedBox& aabb,
+    double& t) {
+  const glm::dvec3& dir = ray.getDirection();
+  const glm::dvec3& origin = ray.getOrigin();
+  const glm::dvec3* min = reinterpret_cast<const glm::dvec3*>(&aabb.minimumX);
+  const glm::dvec3* max = reinterpret_cast<const glm::dvec3*>(&aabb.maximumX);
+  double greatestMin = -std::numeric_limits<double>::max();
+  double smallestMax = std::numeric_limits<double>::max();
+  double tmin = greatestMin;
+  double tmax = smallestMax;
+
+  for (uint32_t i = 0; i < 3; ++i) {
+    if (glm::abs(dir[i]) < Math::Epsilon6) {
+      continue;
+    } else {
+      tmin = ((*min)[i] - origin[i]) / dir[i];
+      tmax = ((*max)[i] - origin[i]) / dir[i];
+    }
+    if (tmin > tmax) {
+      std::swap(tmin, tmax);
+    }
+    greatestMin = glm::max(tmin, greatestMin);
+    smallestMax = glm::min(tmax, smallestMax);
+  }
+  if (smallestMax < 0.0 || greatestMin > smallestMax) {
+    return false;
+  }
+  t = greatestMin < 0.0 ? smallestMax : greatestMin;
+  return true;
+}
+
+std::optional<glm::dvec3>
+IntersectionTests::rayOBB(const Ray& ray, const OrientedBoundingBox& obb) {
+  double t;
+  return rayOBBParametric(ray, obb, t) && t >= 0
+             ? std::make_optional<glm::dvec3>(ray.pointFromDistance(t))
+             : std::nullopt;
+}
+
+bool IntersectionTests::rayOBBParametric(
+    const Ray& ray,
+    const OrientedBoundingBox& obb,
+    double& t) {
+
+  const glm::dmat3x3& inverseHalfAxis = obb.getInverseHalfAxes();
+  glm::dmat4x4 transformation(
+      glm::normalize(glm::dvec4(inverseHalfAxis[0], 0.0)),
```

Comment:
Bit of a nitpick, but swapping the order of the normalize and dvec4 construction would save a multiply and an add (for each column) by normalizing a dvec3 instead of a dvec4.

Reply:
Nah, not a nitpick. Valid optimization. Committed